### PR TITLE
Formatted messages

### PR DESCRIPTION
**Pull request by DanilaFe**
Just  getting started on this. See [FORMATTED](https://github.com/vector-im/hydrogen-web/blob/677bd2984b74aef45ce6151fd135035dc66b2e6b/src/domain/session/room/timeline/FORMATTED.md) for my implementation ideas.

### CODE REVIEW

File: src/domain/session/room/timeline/FORMATTED.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  Then, much like Pandoc AST, we can turn our representation into
+  markdown in the editor, or HTML in the view.
+* As such, we represent formatting by nesting parts
+  (we'd have a `ItalicsPart`, `BoldPart`, etc.)
+* We keep the "inline"/"block" distinction, but only
+  track it as a property, so that we can avoid adding
+  block parts to elements that cannot contain blocks
+  (headers, for instance, cannot contain blocks, but
+  lists -- themselves blocks -- can).
+* When parsing, we may need some sort of "permanent" context:
```
Comment (bwindels) [inline src/domain/session/room/timeline/FORMATTED.md:20]: Indeed, that's how I did this in element web fwiw: https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/editor/deserialize.ts#L94-L99

File: src/platform/web/ui/session/room/timeline/TextMessageView.js [hunk 2]
```diff
@@ -36,18 +36,29 @@ export class TextMessageView extends BaseMessageView {
  * Map from part to function that outputs DOM for the part
  */
 const formatFunction = {
+    header: headerBlock => tag["h" + Math.min(6,headerBlock.level)]({}, renderParts(headerBlock.inlines)),
```
Comment (bwindels) [inline src/platform/web/ui/session/room/timeline/TextMessageView.js:39]: Should the min be on the model, or is your reasoning that it's a HTML limitation to only allow 6 nested headers?

File: src/domain/session/room/timeline/FORMATTED.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  if we're parsing a header, and we are inside 3 layers of other
+  "inline" things (italics, strikethrough, and bold, for example),
+  and we encounter a block, that's still not valid.
+  * Element seems to not care at all about the validity of what
+    it's parsing. Do we assume the HTML is well-formatted, then?
```
Comment (bwindels) [inline src/domain/session/room/timeline/FORMATTED.md:25]: I would think that `DOMParser` will try to wrangle the html into a valid DOM, but could be mistaken.

File: src/platform/web/ui/session/room/timeline/TextMessageView.js (hunk 2 above)
Comment (DanilaFe) [reply src/platform/web/ui/session/room/timeline/TextMessageView.js:39]: Indeed, my assumption was that 6 is an HTML limitation. I think that there's no real reason to enforce the magic number 6 on our "format-independent" data (although we'd probably be hard-pressed to find a source format that does use more than 6 headers)

File: src/domain/session/room/timeline/FORMATTED.md (hunk 3 above)
Comment (DanilaFe) [reply src/domain/session/room/timeline/FORMATTED.md:25]: It does, and so we need not check for validity.

**Comment by DanilaFe** [comment]
Something's up with this pull request; it's showing classes that have been deleted, and duplicate commit. I'm going to close and re-open. Sorry for the noise.

**Pull request closed without merging**